blockchain/pool.go

```diff
@@ -1,630 +1,372 @@
+		if expPeerID, pok := pool.blocks[block.Height]; pok {
+			pool.logger.Error("expected this block from peer", "peer", expPeerID)
+		}
+		return errBadDataFromPeer
 	}
-	peer.numPending++
-}
-
-func (peer *bpPeer) decrPending(recvSize int) {
-	peer.numPending--
-	if peer.numPending == 0 {
-		peer.timeout.Stop()
-	} else {
-		peer.recvMonitor.Update(recvSize)
-		peer.resetTimeout()
+	if b != nil {
+		pool.logger.Error("already have a block for height", "height", block.Height)
+		return errBadDataFromPeer
 	}
-}
-
-func (peer *bpPeer) onTimeout() {
-	peer.pool.mtx.Lock()
-	defer peer.pool.mtx.Unlock()
 
-	err := errors.New("peer did not send us anything")
-	peer.pool.sendError(err, peer.id)
-	peer.logger.Error("SendTimeout", "reason", err, "timeout", peerTimeout)
-	peer.didTimeout = true
+	pool.peers[peerID].blocks[block.Height] = block
+	pool.blocks[block.Height] = peerID
+	pool.numPending--
+	pool.peers[peerID].decrPending(blockSize)
+	pool.logger.Info("added new block", "height", block.Height, "from_peer", peerID, "total", len(pool.blocks))
+	return nil
 }
 
-//-------------------------------------
+func (pool *blockPool) getBlockAndPeerAtHeight(height int64) (bData *blockData, err error) {
+	peerID := pool.blocks[height]
+	peer := pool.peers[peerID]
+	if peer == nil {
+		return nil, errMissingBlocks
+	}
 
-type bpRequester struct {
-	cmn.BaseService
-	pool       *BlockPool
-	height     int64
-	gotBlockCh chan struct{}
-	redoCh     chan p2p.ID //redo may send multitime, add peerId to identify repeat
+	block, ok := peer.blocks[height]
+	if !ok || block == nil {
+		return nil, errMissingBlocks
+	}
 
-	mtx    sync.Mutex
-	peerID p2p.ID
-	block  *types.Block
-}
+	return &blockData{peer: peer, block: block}, nil
 
-func newBPRequester(pool *BlockPool, height int64) *bpRequester {
-	bpr := &bpRequester{
-		pool:       pool,
-		height:     height,
-		gotBlockCh: make(chan struct{}, 1),
-		redoCh:     make(chan p2p.ID, 1),
+}
 
-		peerID: "",
-		block:  nil,
+func (pool *blockPool) getNextTwoBlocks() (first, second *blockData, err error) {
+	first, err = pool.getBlockAndPeerAtHeight(pool.height)
+	second, err2 := pool.getBlockAndPeerAtHeight(pool.height + 1)
+	if err == nil {
+		err = err2
 	}
-	bpr.BaseService = *cmn.NewBaseService(nil, "bpRequester", bpr)
-	return bpr
+	return
 }
 
-func (bpr *bpRequester) OnStart() error {
-	go bpr.requestRoutine()
-	return nil
-}
+// Remove the peers that sent us the first two blocks, blocks are removed by removePeer().
+func (pool *blockPool) invalidateFirstTwoBlocks(err error) {
+	first, err1 := pool.getBlockAndPeerAtHeight(pool.height)
+	second, err2 := pool.getBlockAndPeerAtHeight(pool.height + 1)
 
-// Returns true if the peer matches and block doesn't already exist.
-func (bpr *bpRequester) setBlock(block *types.Block, peerID p2p.ID) bool {
-	bpr.mtx.Lock()
-	if bpr.block != nil || bpr.peerID != peerID {
-		bpr.mtx.Unlock()
-		return false
+	if err1 == nil {
```

Comment:
They shouldn't be non-nil but the logic is in the caller and a bit far away, in the reactor processBlock(). Indeed it will never call this if either blocks are missing. 
At this point invalidateFirstTwoBlocks() is more generic than the caller needs it to be and invalidates the blocks if they exists, does nothing if they don't. I even wrote tests for these cases :)
I could remove the checks and tests if that's what you have in mind (??)